examples/cloud-composer-cicd/cloudbuild.yaml

```diff
@@ -0,0 +1,81 @@
+# Copyright 2018 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+
+steps:
+- name: 'gcr.io/cloud-builders/docker'
+  id: Pull docker cache
+  entrypoint: 'bash'
+  args:
+  - '-c'
+  - |
+   docker pull gcr.io/$PROJECT_ID/airflow-dags-builder:latest || exit 0
+- name: gcr.io/cloud-builders/docker
+  id: Build Airflow DAGs Builder
+  args: [
+      'build',
+      '-t', 'gcr.io/$PROJECT_ID/airflow-dags-builder',
+      '--cache-from', 'gcr.io/$PROJECT_ID/airflow-dags-builder:latest',
+      './'
+    ]
+  dir: ${_DIRECTORY}
+- name: 'gcr.io/$PROJECT_ID/airflow-dags-builder'
+  id: Validation Test
+  # Validate the integrity of the DAG files.
+  entrypoint: python
+  env:
+  - AIRFLOW__CORE__DAGS_FOLDER=/workspace/${_DIRECTORY}/dags
+  args:
+  - -m
+  - unittest
+  - tests/dag_integrity_test.py
+  dir: ${_DIRECTORY}
+- name: 'gcr.io/$PROJECT_ID/airflow-dags-builder'
+  # Actually run some DAGs with test data, this type of test is less common and requires
+  # more setup, creating test data sets, ensuring your task IO is configured
+  # via airflow Variables, etc.
+  # To succeed, these need to complete before the cloudbuild timeout.
+  id: End to End Test
+  entrypoint: bash
+  env:
+  - AIRFLOW__CORE__DAGS_FOLDER=/workspace/${_DIRECTORY}/dags
+  dir: ${_DIRECTORY}
+  # dir: examples/cloud-composer-cicd
+  args:
+  - -c
+  - |
+    pwd
+    ls -l
+    # Init an empty airflow metadata db
+    airflow initdb
+    # Initialise variables to configure airflow DAGs
+    airflow variables -s input gs://test-input-bucket/
+    airflow variables -s output gs://test-output-bucket/
+    # Run a backfill of our DAG for a specific date, add more dags here if required.
+    airflow backfill echo_and_wait --start_date=2020-01-03 --end_date=2020-01-03
+    ret=$?
+    # Cat the task logs so we can see when there was any error
+    find /workspace/airflow/logs/ -type f -name *.log | sort | xargs cat
+    # Fail if the backfill failed.
+    exit $ret
```

Comment:
This should be taken out into it's own script.
This also notably will cause red-herring test failures because it won't handle:
- if the customer has any custom conn_ids
- if the customer is using crypto / fernet key
- if the customer is also deploying custom plugins.

PTAL https://source.cloud.google.com/datapipelines-ci/datapipelines-ci/+/master:composer/cloudbuild/bin/run_tests.sh

Reply:
I don't have access to that repo, permission denied

Reply:
I don't have access to that repo, permission denied

Reply:
I thought it might be more readable in the same file. And since this only tests one DAG this way, it's more of an example than a best practice. This is just to demo how you might override variables before creating and executing a DAG run. 

I definitely agree a more complete solution would be great. But it would be good to have something simple like this to link to until then. We can change to link to that when it's ready.

Reply:
I'd suggest we merge this for now, and then come up with the better solution next. I'd suggest we're much better off having velocity and revising than waiting for perfect.

Reply:
Also, I consider the main benefit to this example to be just the syntax checking, it covers the most immediate need for customers. People tend to do very different things for end to end tests, and most don't, they just see what happens in prod. So it's more of a starting point and helps explain how simple doing things with cloud build is.

Reply:
i opened a PR a while back for the cicd for data pipelines solution that didn't get much traction is very similar to what I'm working on.
You could link to my fork: https://github.com/jaketf/ci-cd-for-data-processing-workflow
I think that handling  plugins / connections is a must even for a simple example.
IMHO cloud build should be like a main function: very small high level readable steps, all implementation details like initdb, variables, etc should be burried in scripts.